lib/scorecalculator.rb

```diff
@@ -0,0 +1,76 @@
+class ScoreCalculator
+  def initialize
+    @sheet = []
+  end
+
+  def input(*args)
+    errors(*args)
+    @sheet << args
+    bonuspoints
+    puts "You scored #{args.sum} for frame #{currentframe}. Your overall score so far is #{currentscore}"
+    specialannouncement
+    reset if completegame?
+  end
+
+  def scoreboard
+    score = 0
+    @sheet.map.with_index do |x, i|
+      puts "Frame #{i + 1}: #{x.flatten} | #{x.flatten.sum} | #{score += x.flatten.sum}"
+    end
+  end
+
+  private
+
+  def errors(*args)
+    raise 'Score in the frame cannot exceed 10!' if currentframe < 9 && args.sum > 10
+    raise "Max of 2 score inputs allowed if you haven't striked/spared in frame 10" if currentframe == 9 && args.length > 2 && ((args[0] + args[1]) < 10)
+    raise "Max of 2 score inputs allowed if you haven't striked/spared!" if args.length > 2 && currentframe < 9
+    raise 'Only 1 score input allowed if it includes a strike' if args.include?(10) && currentframe < 9 && args.length > 1
+    raise "You need to input both scores of the current frame if you don't have a strike" if args.length == 1 && args != [10]
+  end
+
+  def bonuspoints
+    # for continuous strikes
+    if @sheet.length >= 3 && @sheet[currentframe - 3].first == 10 && @sheet[currentframe - 2].first == 10
```

Comment:
these lines of logic are a little long and hard to get your head round - it would probably help with readability to extract some of this into other methods (like whether a frame is a strike, for example).